modules/src/core/ics02_client/client_state.rs

```diff
@@ -1,39 +1,36 @@
 use core::marker::{Send, Sync};
 use core::time::Duration;
 
+use dyn_clone::DynClone;
+use erased_serde::Serialize as ErasedSerialize;
 use ibc_proto::google::protobuf::Any;
-use ibc_proto::protobuf::Protobuf;
-use serde::{Deserialize, Serialize};
+use ibc_proto::protobuf::Protobuf as ErasedProtobuf;
 
-use ibc_proto::ibc::core::client::v1::IdentifiedClientState;
-
-use crate::clients::ics07_tendermint::client_state;
-use crate::clients::ics07_tendermint::client_state::ClientState as TmClientState;
 use crate::core::ics02_client::client_type::ClientType;
 use crate::core::ics02_client::error::Error;
-use crate::core::ics02_client::trust_threshold::TrustThreshold;
-use crate::core::ics24_host::error::ValidationError;
-use crate::core::ics24_host::identifier::{ChainId, ClientId};
-#[cfg(any(test, feature = "mocks"))]
-use crate::mock::client_state::MockClientState;
+use crate::core::ics24_host::identifier::ChainId;
+use crate::dynamic_typing::AsAny;
 use crate::prelude::*;
 use crate::Height;
 
-pub const TENDERMINT_CLIENT_STATE_TYPE_URL: &str = "/ibc.lightclients.tendermint.v1.ClientState";
-pub const MOCK_CLIENT_STATE_TYPE_URL: &str = "/ibc.mock.ClientState";
-
-pub trait ClientState: core::fmt::Debug + Send + Sync {
-    /// Client-specific options for upgrading the client
-    type UpgradeOptions;
-
+pub trait ClientState:
+    AsAny
+    + sealed::ErasedPartialEqClientState
+    + DynClone
+    + ErasedSerialize
+    + ErasedProtobuf<Any, Error = Error>
+    + core::fmt::Debug
+    + Send
+    + Sync
+{
     /// Return the chain identifier which this client is serving (i.e., the client is verifying
     /// consensus states from this chain).
     fn chain_id(&self) -> ChainId;
 
     /// Type of client associated with this state (eg. Tendermint)
     fn client_type(&self) -> ClientType;
 
-    /// Latest height of consensus state
+    /// Latest height of client state
```

Comment:
I think the previous comment was actually correct. The client state doesn't get updated on every header; the consensus state does. I think this comment comes from the ibc-go implementation (e.g. [here](https://github.com/cosmos/ibc-go/blob/0a887e47318c21dc4668c1056ceae7a7e795a28c/modules/core/02-client/keeper/events.go#L24) the consensus height is being set with `ClientState.latest_height()`.

We should probably find a better name when we refactor the context traits.

Reply:
I found `Latest height of consensus state` ambiguous because there would be multiple consensus states stored for a single light client (or client state). I think it should be something like `height of most recently installed header` or `Latest height the client was updated to` (this is what the [proto def has](https://github.com/informalsystems/ibc-rs/blob/master/proto/src/prost/ibc.lightclients.tendermint.v1.rs#L22)). WDYT?

Reply:
Good point, "Latest height the client was updated to" sounds good!

Reply:
https://github.com/informalsystems/ibc-rs/pull/2559/commits/8d6bfcd42970ed86daf11838ae8ff771f4785841